File: swift/BCLifeHash/Sources/BCLifeHash/CellGrid.swift
final class CellGrid {
    let grid: Grid<Bool>

    init(width: Int, height: Int) {
        self.grid = Grid(width: width, height: height, defaultValue: false)
    }

    private static func isAliveInNextGeneration(_ currentAlive: Bool, _ neighborsCount: Int) -> Bool {
        if currentAlive {
            return neighborsCount == 2 || neighborsCount == 3
        }
        return neighborsCount == 3
    }

    private func countNeighbors(_ px: Int, _ py: Int) -> Int {
        var total = 0
        grid.forNeighborhood(px, py) { ox, oy, nx, ny in
            if ox == 0 && oy == 0 {
                return
            }
            if grid[nx, ny] {
                total += 1
            }
        }
        return total
    }

    func data() -> [UInt8] {
        let aggregator = BitAggregator()
        grid.forEachCell { x, y in
            aggregator.append(grid[x, y])
        }
        return aggregator.valueData()
    }

    func setData(_ data: [UInt8]) {
        precondition(grid.width * grid.height == data.count * 8)
        let enumerator = BitEnumerator(data: data)
        var i = 0
        enumerator.forAll { bit in
            grid.storage[i] = bit
            i += 1
        }
        precondition(i == grid.storage.count)
    }

    func nextGeneration(
        currentChangeGrid: ChangeGrid,
        nextCellGrid: CellGrid,
        nextChangeGrid: ChangeGrid
    ) {
        nextCellGrid.grid.setAll(false)
        nextChangeGrid.grid.setAll(false)

        let width = grid.width
        let height = grid.height

        for y in 0..<height {
            for x in 0..<width {
                let currentAlive = grid[x, y]
                if currentChangeGrid.grid[x, y] {
                    let neighborsCount = countNeighbors(x, y)
                    let nextAlive = Self.isAliveInNextGeneration(currentAlive, neighborsCount)
                    if nextAlive {
                        nextCellGrid.grid[x, y] = true
                    }
                    if currentAlive != nextAlive {
                        nextChangeGrid.setChanged(x, y)
                    }
                } else {
                    nextCellGrid.grid[x, y] = currentAlive
                }
            }
        }
    }
}

File: kotlin/bc-lifehash/src/main/kotlin/com/blockchaincommons/bclifehash/CellGrid.kt
package com.blockchaincommons.bclifehash

/**
 * A boolean grid representing alive/dead cells for Conway's Game of Life.
 */
internal class CellGrid(width: Int, height: Int) {
    val grid = Grid(width, height, false)

    private fun isAliveInNextGeneration(currentAlive: Boolean, neighborsCount: Int): Boolean =
        if (currentAlive) {
            neighborsCount == 2 || neighborsCount == 3
        } else {
            neighborsCount == 3
        }

    private fun countNeighbors(px: Int, py: Int): Int {
        var total = 0
        grid.forNeighborhood(px, py) { offsetX, offsetY, neighborX, neighborY ->
            if (offsetX == 0 && offsetY == 0) {
                return@forNeighborhood
            }
            if (grid.getValue(neighborX, neighborY)) {
                total += 1
            }
        }
        return total
    }

    fun toByteArray(): ByteArray {
        val aggregator = BitAggregator()
        grid.forEach { x, y ->
            aggregator.append(grid.getValue(x, y))
        }
        return aggregator.toByteArray()
    }

    fun loadFrom(data: ByteArray) {
        val enumerator = BitEnumerator(data.copyOf())
        var index = 0
        enumerator.forEach { bit ->
            grid.storage[index] = bit
            index += 1
        }
        check(index == grid.storage.size) { "CellGrid data size mismatch" }
    }

    fun nextGeneration(
        currentChangeGrid: ChangeGrid,
        nextCellGrid: CellGrid,
        nextChangeGrid: ChangeGrid,
    ) {
        nextCellGrid.grid.fill(false)
        nextChangeGrid.grid.fill(false)

        val width = grid.width
        val height = grid.height
        for (y in 0 until height) {
            for (x in 0 until width) {
                val currentAlive = grid.getValue(x, y)
                if (currentChangeGrid.grid.getValue(x, y)) {
                    val neighborsCount = countNeighbors(x, y)
                    val nextAlive = isAliveInNextGeneration(currentAlive, neighborsCount)
                    if (nextAlive) {
                        nextCellGrid.grid.setValue(true, x, y)
                    }
                    if (currentAlive != nextAlive) {
                        nextChangeGrid.setChanged(x, y)
                    }
                } else {
                    nextCellGrid.grid.setValue(currentAlive, x, y)
                }
            }
        }
    }
}
